// antd-mini/es/VTabs/VTabItem/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var props_1 = require("./props");
var context_1 = require("../context");
var console_1 = require("../../_util/console");
var tools_1 = require("../../_util/tools");
var component2 = my.canIUse('component2');
Component({
    props: props_1.VTabItemDefaultProps,
    data: {
        component2: component2,
    },
    didMount: function () {
        this._getTabInfo();
    },
    didUpdate: function () {
        this._getTabInfo();
    },
    didUnmount: function () {
        delete context_1.getTabArray[this.$id];
    },
    methods: {
        _tabError: function (tab) {
            // 检测 tab-item 中的 tab 属性值正确性
            if (!tab || typeof tab !== 'object') {
                console_1.log.error('VTabItem', "tab \u4E2D\u5B58\u6709\u6570\u636E\u7C7B\u578B\u9519\u8BEF\uFF0C\u5E94\u4E3A object\uFF0C\u5F53\u524D\u7684 tab \u5C5E\u6027\u503C\u7C7B\u578B\u4E3A ".concat(typeof tab, "\uFF0C\u503C\uFF1A").concat(tab));
            }
            else if (!(tab === null || tab === void 0 ? void 0 : tab.title)) {
                console_1.log.error('VTabItem', "tab \u7684\u503C\u4E2D\u7F3A\u5C11\u5173\u952E title \u503C\uFF0C\u5F53\u524D\u7684 tab \u503C\u4E3A ".concat(JSON.stringify(tab)));
            }
            if ((tab === null || tab === void 0 ? void 0 : tab.count) && typeof (tab === null || tab === void 0 ? void 0 : tab.count) !== 'number') {
                console_1.log.error('VTabItem', "tab \u4E2D\u7684 count \u7C7B\u578B\u4E0D\u5339\u914D\uFF0C\u5F53\u524D\u7684 count \u503C\u4E3A ".concat(tab === null || tab === void 0 ? void 0 : tab.count, "\uFF0C \u7C7B\u578B\u4E3A ").concat(typeof (tab === null || tab === void 0 ? void 0 : tab.count)));
            }
        },
        _getTabInfo: function () {
            // 获取每个 tab-item 中的 tab 值
            // 如果是 didMount 时对 uid 的值做累加处理
            var tab = this.props.tab;
            this._tabError(tab);
            context_1.getTabArray[this.$id] = tab;
            setTimeout(function () {
                context_1.componentContext.update((0, tools_1.objectValues)(context_1.getTabArray));
            }, 0);
        },
    },
});
