// antd-mini/es/RadioGroup/RadioItem/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var props_1 = require("./props");
var context_1 = require("../context");
Component({
    props: props_1.RadioItemDefaultProps,
    data: {
        _checked: false,
        _disabled: false,
    },
    didMount: function () {
        var _this = this;
        var uid = this.props.uid;
        var key = "".concat(this.$page.$id, "-").concat(uid);
        context_1.componentDisabled.onUpdate(key, (this.disabledListener = function (d) {
            _this.setData({
                _disabled: _this.props.disabled || d,
            });
        }));
        context_1.componentValue.onUpdate(key, (this.checkedListener = function (v) {
            if (v === _this.props.value) {
                _this.setData({
                    _checked: true,
                });
            }
            else {
                _this.setData({
                    _checked: false,
                });
            }
        }));
    },
    didUnmount: function () {
        var uid = this.props.uid;
        var key = "".concat(this.$page.$id, "-").concat(uid);
        context_1.componentValue.offUpdate(key, this.checkedListener);
        context_1.componentDisabled.offUpdate(key, this.disabledListener);
    },
    methods: {
        onItemChange: function (v) {
            var uid = this.props.uid;
            var key = "".concat(this.$page.$id, "-").concat(uid);
            var value = v.detail.value;
            if (value) {
                context_1.componentContext.update(key, this.props.value);
            }
        },
    },
});
