// antd-mini/es/Loading/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var props_1 = require("./props");
Component({
    props: props_1.LoadingDefaultProps,
    data: {
        _loading: false,
    },
    didMount: function () {
        this.setLoading();
    },
    didUpdate: function (prevProps) {
        if (prevProps.loading !== this.props.loading) {
            if (this.timeout) {
                clearTimeout(this.timeout);
            }
            this.setLoading();
        }
    },
    methods: {
        setLoading: function () {
            var _this = this;
            var _a = this.props, delay = _a.delay, loading = _a.loading;
            var realDelay = isNaN(delay) ? 0 : delay;
            if (loading && !this.data._loading) {
                this.timeout = setTimeout(function () {
                    _this.setData({
                        _loading: true,
                    });
                }, realDelay);
            }
            else if (!loading && this.data._loading) {
                this.setData({
                    _loading: false,
                });
            }
        },
    },
});
