// antd-mini/es/Steps/StepItem/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var props_1 = require("./props");
var context_1 = require("../context");
Component({
    props: props_1.StepItemDefaultProps,
    data: {
        _index: 0,
        _direction: 'horizontal',
    },
    didMount: function () {
        var _this = this;
        var uid = this.props.uid;
        var key = "".concat(this.$page.$id, "-").concat(uid);
        var setItemIndex = function (index) { return _this.setData({
            _index: index,
        }); };
        var setItemDirection = function (direction) { return _this.setData({
            _direction: direction,
        }); };
        context_1.context.addItem(key, "".concat(this.$id), {
            setItemIndex: setItemIndex,
            setItemDirection: setItemDirection,
        });
    },
    didUnmount: function () {
        var key = "".concat(this.$page.$id, "-").concat(this.props.uid);
        context_1.context.removeItem(key, "".concat(this.$id));
    },
    methods: {},
});
