// antd-mini/es/Picker/props.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PickerDefaultProps = void 0;

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

var PickerDefaultProps = {
  controlled: false,
  okText: '确定',
  dismissText: '取消',
  disabled: false,
  maskClosable: false,
  value: '',
  data: [],
  placeholder: '请选择',
  onFormat: function onFormat(value, column) {
    if (column instanceof Array) {
      return column.filter(function (c) {
        return c !== undefined;
      }).map(function (c) {
        if (_typeof(c) === 'object') {
          return c.label;
        }

        return c;
      }).join('-');
    }

    return column && column.label || column;
  }
};
exports.PickerDefaultProps = PickerDefaultProps;
