// antd-mini/es/Terms/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var fmtEvent_1 = tslib_1.__importDefault(require("../_util/fmtEvent"));
Component({
    props: props_1.TermsDefaultProps,
    data: {
        disabled: false,
    },
    didMount: function () {
        this.updateDisabled(false);
    },
    methods: {
        updateDisabled: function (v) {
            if (this.props.hasCheckbox) {
                this.setData({
                    disabled: !v,
                });
            }
        },
        onChange: function (v) {
            var _a = this.props, onChange = _a.onChange, _getCurrentField = _a._getCurrentField;
            this.updateDisabled(v);
            if (onChange) {
                var event = (0, fmtEvent_1.default)(this.props);
                if (_getCurrentField) {
                    return onChange.call(this.props, v, event);
                }
                else {
                    return onChange(v, event);
                }
            }
        },
        onMainBtnTap: function () {
            if (typeof this.props.onMainBtnTap === 'function') {
                this.props.onMainBtnTap((0, fmtEvent_1.default)(this.props));
            }
        },
        onSubBtnTap: function () {
            if (typeof this.props.onSubBtnTap === 'function') {
                this.props.onSubBtnTap((0, fmtEvent_1.default)(this.props));
            }
        },
    },
});
