// antd-mini/es/Popup/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var props_1 = require("./props");
Component({
    props: props_1.PopupDefaultProps,
    data: {
        supportSjs: my.canIUse('sjs.event'),
        closing: false,
    },
    didUpdate: function (prevProps) {
        var _a = this.props, visible = _a.visible, duration = _a.duration, animation = _a.animation;
        if (prevProps.visible && !visible) {
            if (animation && duration > 0) {
                this.setData({ closing: true });
            }
        }
    },
    methods: {
        onMaskClose: function () {
            var closing = this.data.closing;
            if (closing) {
                return;
            }
            var maskClosable = this.props.maskClosable;
            if (maskClosable) {
                this.onClose();
            }
        },
        onClose: function () {
            var onClose = this.props.onClose;
            if (onClose) {
                onClose();
            }
        },
        onAnimationEnd: function () {
            var closing = this.data.closing;
            if (closing) {
                this.setData({ closing: false });
            }
        },
    },
});
