// antd-mini/es/_util/promisify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadFile = exports.previewImage = exports.getFileInfo = exports.uploadFile = exports.chooseFileFromDisk = exports.chooseVideo = exports.chooseImage = void 0;
var tslib_1 = require("tslib");
/*
 *@Description: 把所有my.xx的方法变成promise
 */
function promisifyMy(name) {
    return function (options) {
        return new Promise(function (resolve, reject) {
            my[name](tslib_1.__assign(tslib_1.__assign({}, options), { success: resolve, fail: reject }));
        });
    };
}
/*
 *@Description: 把所有my.call(xx)的方法变成promise
 */
function promisifyCall(name) {
    return function (options) {
        return new Promise(function (resolve) {
            my.call(name, tslib_1.__assign({}, options), resolve);
        });
    };
}
exports.chooseImage = promisifyMy('chooseImage');
exports.chooseVideo = promisifyMy('chooseVideo');
exports.chooseFileFromDisk = promisifyMy('chooseFileFromDisk');
exports.uploadFile = promisifyMy('uploadFile');
exports.getFileInfo = promisifyMy('getFileInfo');
exports.previewImage = promisifyMy('previewImage');
exports.downloadFile = promisifyMy('downloadFile');
