// antd-mini/es/TabBar/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/no-explicit-any */
var props_1 = require("./props");
Component({
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    props: props_1.TabBarDefaultProps,
    data: {
        activeIndex: undefined,
    },
    didMount: function () {
        var activeKey = this.props.activeKey;
        this.setData({
            activeIndex: activeKey,
        });
    },
    didUpdate: function () {
        var activeKey = this.props.activeKey;
        var activeIndex = this.data.activeIndex;
        if (activeIndex === activeKey)
            return;
        this.setData({
            activeIndex: activeKey,
        });
    },
    methods: {
        onChange: function (e) {
            var index = e.target.dataset.index;
            var activeIndex = this.data.activeIndex;
            if (activeIndex === index)
                return;
            this.setData({
                activeIndex: index,
            });
            this.props.onChange && this.props.onChange(this.props[index], index);
        },
    },
});
