// antd-mini/es/Checklist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var controlled_1 = tslib_1.__importDefault(require("../mixins/controlled"));
var fmtEvent_1 = tslib_1.__importDefault(require("../_util/fmtEvent"));
Component({
    props: props_1.ChecklistDefaultProps,
    mixins: [(0, controlled_1.default)()],
    didMount: function () {
        var _a = this.props, multiple = _a.multiple, value = _a.value;
        if (multiple && !Array.isArray(value)) {
            this.setData({
                cValue: []
            });
        }
    },
    methods: {
        onChange: function (value) {
            var _a;
            var _b = this.props, multiple = _b.multiple, options = _b.options;
            var items = null;
            if (multiple && Array.isArray(value)) {
                items = value.map(function (v) {
                    var _a;
                    return (_a = options.filter(function (o) { return o.value === v; })) === null || _a === void 0 ? void 0 : _a[0];
                });
            }
            else {
                items = (_a = options.filter(function (o) { return o.value === value; })) === null || _a === void 0 ? void 0 : _a[0];
            }
            this.cOnChange(value, items, (0, fmtEvent_1.default)(this.props));
        }
    },
});
