// antd-mini/es/CascaderPicker/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var props_1 = require("./props");
var form_1 = tslib_1.__importDefault(require("../mixins/form"));
var fmtEvent_1 = tslib_1.__importDefault(require("../_util/fmtEvent"));
var fast_deep_equal_1 = tslib_1.__importDefault(require("fast-deep-equal"));
Component({
    mixins: [(0, form_1.default)({ trigger: 'onOk' })],
    _visible: false,
    props: props_1.CascaderDefaultProps,
    data: function () {
        return {
            currentValue: [],
            columns: [],
            cValue: null,
        };
    },
    didMount: function () {
        var value = this.props.value;
        var columns = this.getterColumns(value);
        // 首次无需校验value有效性，onOk时会校验
        this.setData({ columns: columns, cValue: value });
    },
    didUpdate: function (prevProps) {
        var _a = this.props, value = _a.value, options = _a.options;
        var _b = this.data, columns = _b.columns, currentValue = _b.currentValue;
        // onTriggerPicker展开时会自动重置数据，此处只有展开状态下才需要重置columns和currentValue，否则只需设置cValue
        if (this._visible) {
            if (options !== prevProps.options) {
                var newData = {};
                if (!(0, fast_deep_equal_1.default)(value, prevProps.value)) {
                    var newColumns = this.getterColumns(value);
                    newData.columns = newColumns;
                    var currentValue_1 = this.getValidValue(value, newColumns);
                    newData.currentValue = currentValue_1;
                    newData.cValue = value;
                }
                else {
                    var newColumns = this.getterColumns(currentValue);
                    newData.columns = newColumns;
                }
                this.setData(newData);
            }
            else {
                if (!(0, fast_deep_equal_1.default)(value, prevProps.value)) {
                    var currentValue_2 = this.getValidValue(value, columns);
                    this.setData({ cValue: value, currentValue: currentValue_2 });
                }
            }
        }
        else {
            if (!(0, fast_deep_equal_1.default)(value, prevProps.value)) {
                this.setData({
                    cValue: value,
                    currentValue: this.getValidValue(value, columns),
                });
            }
        }
    },
    methods: {
        getterColumns: function (value) {
            var getColumns = function (options, value, columns) {
                var _a;
                if (columns === void 0) { columns = []; }
                columns.push(options.map(function (v) { return ({ value: v.value, label: v.label }); }));
                var currentOption = options.find(function (v) { return v.value === (value === null || value === void 0 ? void 0 : value[columns.length - 1]); }) ||
                    options[0];
                if (((_a = currentOption === null || currentOption === void 0 ? void 0 : currentOption.children) === null || _a === void 0 ? void 0 : _a.length) > 0) {
                    return getColumns(currentOption.children, value, columns);
                }
                return columns;
            };
            var options = this.props.options;
            return getColumns(options, value);
        },
        // 获取有效value，若从x项开始在columns里找不到，则从此项开始都选第一条
        getValidValue: function (value, columns) {
            var result = [];
            var _loop_1 = function (i) {
                if (!columns[i].some(function (v) { return v.value === (value === null || value === void 0 ? void 0 : value[i]); })) {
                    result.push.apply(result, columns.slice(i).map(function (v) { return v[0].value; }));
                    return "break";
                }
                else {
                    result[i] = value[i];
                }
            };
            for (var i = 0; i < columns.length; i++) {
                var state_1 = _loop_1(i);
                if (state_1 === "break")
                    break;
            }
            return result;
        },
        getOptionByValue: function (value) {
            var _a;
            var options = this.props.options;
            if (!((value === null || value === void 0 ? void 0 : value.length) > 0))
                return null;
            var result = [];
            var item = options.find(function (v) { return v.value === value[0]; });
            var _loop_2 = function (i) {
                if (!item) {
                    return { value: null };
                }
                result.push({
                    value: item.value,
                    label: item.label,
                });
                item = (_a = item.children) === null || _a === void 0 ? void 0 : _a.find(function (v) { return v.value === value[i + 1]; });
            };
            for (var i = 0; i < value.length; i++) {
                var state_2 = _loop_2(i);
                if (typeof state_2 === "object")
                    return state_2.value;
            }
            return result;
        },
        onChange: function (selectedValue) {
            var onPickerChange = this.props.onPickerChange;
            var columns = this.data.columns;
            var newColumns = this.getterColumns(selectedValue);
            // columns没变化说明selectedValue在范围内，无需重置
            var newData = {};
            if (!(0, fast_deep_equal_1.default)(columns, newColumns)) {
                selectedValue = this.getValidValue(selectedValue, newColumns);
                newData.columns = newColumns;
            }
            newData.currentValue = selectedValue;
            this.setData(newData);
            if (onPickerChange) {
                onPickerChange(selectedValue, this.getOptionByValue(selectedValue), (0, fmtEvent_1.default)(this.props));
            }
        },
        onOk: function () {
            return tslib_1.__awaiter(this, void 0, void 0, function () {
                var _a, currentValue, columns, _b, onOk, onBeforeOk, validValue, isContinue;
                return tslib_1.__generator(this, function (_c) {
                    switch (_c.label) {
                        case 0:
                            _a = this.data, currentValue = _a.currentValue, columns = _a.columns;
                            _b = this.props, onOk = _b.onOk, onBeforeOk = _b.onBeforeOk;
                            validValue = this.getValidValue(currentValue, columns);
                            if (!onBeforeOk) return [3 /*break*/, 2];
                            return [4 /*yield*/, onBeforeOk(validValue, this.getOptionByValue(validValue), (0, fmtEvent_1.default)(this.props))];
                        case 1:
                            isContinue = _c.sent();
                            if (!isContinue) {
                                return [2 /*return*/];
                            }
                            _c.label = 2;
                        case 2:
                            this.setData({ cValue: validValue });
                            if (onOk) {
                                onOk(validValue, this.getOptionByValue(validValue), (0, fmtEvent_1.default)(this.props));
                            }
                            return [2 /*return*/];
                    }
                });
            });
        },
        onTriggerPicker: function (visible) {
            var onTriggerPicker = this.props.onTriggerPicker;
            var _a = this.data, cValue = _a.cValue, columns = _a.columns;
            this._visible = visible;
            if (visible) {
                var newColumns = this.getterColumns(cValue);
                var currentValue = this.getValidValue(cValue, newColumns);
                var newData = { currentValue: currentValue };
                if (!(0, fast_deep_equal_1.default)(columns, newColumns)) {
                    newData.columns = newColumns;
                }
                this.setData(newData);
            }
            if (onTriggerPicker) {
                onTriggerPicker(visible, (0, fmtEvent_1.default)(this.props));
            }
        },
        onFormat: function () {
            var cValue = this.data.cValue;
            return this.props.onFormat.call(this, cValue, this.getOptionByValue(cValue));
        },
        onDismiss: function (e) {
            var onDismiss = this.props.onDismiss;
            if (onDismiss) {
                onDismiss((0, fmtEvent_1.default)(this.props, e));
            }
        },
    },
});
